### DISCUSSION TITLE
Building GOCART in UFS with spack-stack: cannot find esmf.mod

### DISCUSSION THREAD

**Discussion by benjamin-cash** [GEOS-ESM/GOCART#174] [Q&A]
Hello,

Hopefully this is the right place to ask this question. :) Like the title says, I am trying to build GOCART as part of the ufs, using spack-stack, and I am running into the following error:

/work2/02441/bcash/frontera/ufs-weather-model/develop/GOCART/ESMF/Shared/Chem_AeroGeneric.F90:18:7:

   18 |    use ESMF
      |       1
Fatal Error: Cannot open module file ‘esmf.mod’ for reading at (1): No such file or directory

I can build the model in many other configurations without issue, it is specifically when I try to build with GOCART that this issue arises. Do I need to do something beyond loading the esmf module to have it found by GOCART? Thanks!

@climbfuji - Just tagging you in since you were the first one to run into this.

**Comment by benjamin-cash** [comment]
@rsdunlapiv - Since I see you come up here, any thoughts on what could be happening here?

**Reply by rsdunlapiv** [reply]
@benjamin-cash it looks like an issue with the UFS build system using spack-stack. I think this should go to whomever is maintaining the spack-stack and the UFS build system itself.

**Comment by mathomp4** [comment]
@benjamin-cash What version of GOCART are you trying to use? I want to say Spack and GOCART might only work in `develop` at this point. If so, I might need to backport some CMake changes into GOCART 2.0.

For example, these PRs:

https://github.com/GEOS-ESM/GOCART/pull/115
https://github.com/GEOS-ESM/GOCART/pull/126

are only in `develop` I believe. It wouldn't be too hard I think to backport all the changes needed.

**Reply by benjamin-cash** [reply]
Hi @mathomp4 - I am using the version of GOCART currently being pointed to by ufs-weather-model develop. The hash is  a342b1b.

**Reply by benjamin-cash** [reply]
Wiping out my current GOCART and cloning into develop does not seem to have fixed it, my build failed in the same place. 

One thing I noticed that seems to be related is that build/GOCART/ESMF/Shared/CMakeFiles/Chem_Shared2G.dir/flags.make does not include `-I/work2/02441/bcash/frontera/spack-stack/envs/ufs-weather-model.frontera.gnu/install/gcc/9.1.0/esmf-8.3.0b09-ggdjsv3/include`, and on a quick check it seems the other flags.make files in build/GOCART/ESMF do.

**Reply by mathomp4** [reply]
@benjamin-cash can you try editing `ESMF/Shared/CMakeLists.txt` and do this:
```cmake
esma_add_library(${this}
  SRCS ${srcs}
  DEPENDENCIES MAPL esmf)
```
where I added the `esmf` (on line 8)?

I guess in GEOS we get `esmf` via `MAPL` but maybe the dependency isn't working its way down in UFS for...reasons?

**Reply by benjamin-cash** [reply]
@mathomp4 - When I do that I got farther, but then it fails with:
/opt/apps/gcc/9.1.0/lib/gcc/x86_64-pc-linux-gnu/9.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: cannot find -lESMF
16351 collect2: error: ld returned 1 exit status

**Reply by benjamin-cash** [reply]
@mathomp4 - One more update. I went into CMakeFiles/ufs_model.dir/link.txt and removed -lESMF manually and reran make. At that point it completed and produced an executable.

**Reply by mathomp4** [reply]
@benjamin-cash Hmm. How did you get `-lESMF`? That's a weird thing to get. Did you do a complete new CMake with a fresh build dir?

**Reply by benjamin-cash** [reply]
@mathomp4 - In this attempt yes, I started with a clean install of everything, made the various changes, and reran cmake. Although I think I got this same error when I tried restarting the build after making updates as well.

**Reply by mathomp4** [reply]
hmm. Okay. I have a thought. I already made a PR (#177), but I don't think it's enough. Give me a bit to ponder...

**Reply by benjamin-cash** [reply]
It only takes a couple of minutes to test what is in that branch - I'll tell you the result in a bit.

**Reply by mathomp4** [reply]
@benjamin-cash Couple new tests for you. First, you could try the branch in #177, but if that doesn't work, maybe try editing the base `CMakeLists.txt` and where you see:
```cmake
  if (NOT ESMF_FOUND)
    find_package (ESMF REQUIRED)
  endif ()
```
add:
```cmake
  if (NOT TARGET esmf)
    add_library(esmf ALIAS ESMF)
  endif ()
```
right after?

**Reply by benjamin-cash** [reply]
First test of #177 failed with the same /opt/apps/gcc/8.3.0/bin/ld: cannot find -lESMF error. Second test failed as well.

**Reply by mathomp4** [reply]
Well I am baffled. We might need to try this ourselves. Are you running on Orion? If so, we could get @weiyuan-jiang to do a test build. In the end, we might need to do a CMake deep dive to figure out what is trying to get you `-lESMF` since, well, nothing should be providing that or asking for that. Baselibs has definitely never expressly built `libESMF.a/.so`

Since it is spack-stack, I'm wondering if somehow the wrong `FindESMF.cmake` is being used? I'm not sure this would cause this but...

@climbfuji, do you know what version of `FindESMF.cmake` might be being used here? We used to supply one in ESMA_cmake, but even that old one wouldn't look for `libESMF` but `libesmf` as it should. And now we just use the one directly from the ESMF build itself.

**Reply by benjamin-cash** [reply]
This is on Frontera, although I do have Orion access so I could also try it there.

**Reply by mathomp4** [reply]
@benjamin-cash If you can, maybe try that. No one here at GMAO has access to Frontera so we'd have to rely on you or @climbfuji for help.

**Reply by mathomp4** [reply]
Also this:

```
CMakeFiles/ufs_model.dir/link.txt
```

makes me think figuring this out will be "fun". I mean, the CMake for that seems to be:
```cmake
###############################################################################
### UFS executable
###############################################################################
add_executable(ufs_model driver/UFS.F90)
add_dependencies(ufs_model ufs)
target_link_libraries(ufs_model ufs esmf w3emc::w3emc_d)
```
which means a nice long trek through the CMake.

And of course ufs-weather-model has five `FindESMF.cmake` that I know of. And maybe spack-stack is pulling in more? 

Are you able to try building a "GOCART-less" version of UFS? Maybe if removing GOCART fixes this, we know it's somehow on our end...

**Reply by benjamin-cash** [reply]
Would you like access? I have a startup allocation on Frontera specifically to enable porting UFS there and benchmarking the code. I'd be happy to add you or someone else.

**Reply by benjamin-cash** [reply]
@mathomp4 - It is definitely related to GOCART. I've built the ATM, S2S, and S2SW apps without hitting this snag. It comes up when I try to build S2SWA and GOCART.

**Reply by benjamin-cash** [reply]
@mathomp4 - Just poking my head in to see if this ever got resolved.

**Reply by mathomp4** [reply]
@benjamin-cash I've built all of GEOSgcm inside of spack so this should be doable. Have you tried GOCART v2.1.0? 

If that still doesn't work, you could try changing this chunk of code in the `CMakeLists.txt`:
```cmake
  if (NOT ESMF_FOUND)
    find_package (ESMF REQUIRED)
  endif ()
```
to:
```cmake
  if (NOT TARGET esmf)
    if (DEFINED ENV{esmf_ROOT})
      message (STATUS "Found esmf_ROOT in environment: $ENV{esmf_ROOT}")
      find_path(ESMF_CMAKE_PATH FindESMF.cmake HINTS "$ENV{esmf_ROOT}/cmake")
      message (STATUS "Found FindESMF.cmake in: ${ESMF_CMAKE_PATH}")
      if (ESMF_CMAKE_PATH)
        message (STATUS "Appending to CMAKE_PREFIX_PATH: ${ESMF_CMAKE_PATH}")
        list (APPEND CMAKE_MODULE_PATH ${ESMF_CMAKE_PATH})
      endif ()
    endif ()

    find_package(ESMF MODULE REQUIRED)

    # ESMF as used in MAPL requires MPI
    # NOTE: This looks odd because some versions of FindESMF.cmake out in the
    #       world provide an "esmf" target while others provide "ESMF". So we
    #       need this ugliness to support both.
    if (TARGET esmf)
      target_link_libraries(esmf INTERFACE MPI::MPI_Fortran)
    else()
      target_link_libraries(ESMF INTERFACE MPI::MPI_Fortran)
      # MAPL and GEOS use lowercase target due to historical reasons but
      # the latest FindESMF.cmake file from ESMF produces an ESMF target.
      add_library(esmf ALIAS ESMF)
    endif()
  endif ()
```

That bit of horribleness is what I came up with in my current testing of trying `spack install geosgcm`

**Comment by climbfuji** [comment]
@mathomp4 @benjamin-cash I think I can spend some time this afternoon trying to decipher this.

**Comment by benjamin-cash** [comment]
It looks like I can't test on Orion at this point. The installation of spack-stack there does not include w3emc so the model build fails, and when I tried to build spac-stack myself it died almost immediately with an error about the C compiler and "github-actions-v0.1 raised ValueError: source is not trusted". So it will have to wait until someone updates that on Orion or @climbfuji untangles it.
